Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -2,119 +2,300 @@
+require "azure/storage"
+require 'json' # for registry content
+require "securerandom" # for generating uuid.
 
-# Reads events from Azure Blobs
-class LogStash::Inputs::Azureblob < LogStash::Inputs::Base
+# Registry item to coordinate between mulitple clients
+class LogStash::Inputs::RegistryItem
+  attr_accessor :file_path, :etag, :offset, :reader
 
+  # Allow json serialization.
+  def as_json(options={})
+      {
+          file_path: @file_path,
+          etag: @etag,
+          reader: @reader,
+          offset: @offset
+      }
+  end # as_json
+
+  def to_json(*options)
+      as_json(*options).to_json(*options)
+  end # to_json
+
+  def initialize(file_path, etag, reader, offset = 0)
+      @file_path = file_path
+      @etag = etag
+      @reader = reader
+      @offset = offset
+  end # initialize
+end # class RegistryItem
+
+
+# Logstash input plugin for Azure Blobs
+#
+# This logstash plugin gathers data from Microsoft Azure Blobs
+class LogStash::Inputs::LogstashInputAzureblob < LogStash::Inputs::Base
   config_name "azureblob"
-  milestone 1
-  
+
+  # If undefined, Logstash will complain, even if codec is unused.
   default :codec, "json_lines"
-  
+
+  # Set the account name for the azure storage account.
   config :storage_account_name, :validate => :string
+  
+  # Set the key to access the storage account.
   config :storage_access_key, :validate => :string
   
+  # Set the container of the blobs.
   config :container, :validate => :string
-  config :sleep_time, :validate => :number, :default => 10
-  config :endpoint, :validate => :string, :default => "core.windows.net"
-  config :backupmode, :validate => :boolean, :default => false 
   
-  def initialize(*args)
-    super(*args)
-  end # def initialize
+  # Set the endpoint for the blobs.
+  #
+  # The default, `core.windows.net` targets the public azure.
+  config :endpoint, :validate => :string, :default => 'core.windows.net'
+
+  # Set the value of using backup mode.
+  config :backupmode, :validate => :boolean, :default => false, :deprecated => true, :obsolete => 'This option is obsoleted and the settings will be ignored.'
+  
+  # Set the value for the registry file.
+  #
+  # The default, `data/registry`, is used to coordinate readings for various instances of the clients.
+  config :registry_path, :validate => :string, :default => 'data/registry'
   
+  # Set how frequently messages should be sent.
+  #
+  # The default, `30`, means trigger a reading for the log every 30 seconds.
+  config :interval, :validate => :number, :default => 30
+
   public
   def register
-    Azure.configure do |config|
-      config.storage_account_name = @storage_account_name
-      config.storage_access_key = @storage_access_key
-      config.storage_blob_host = "https://#{@storage_account_name}.blob.#{@endpoint}"
-    end
-    @azure_blob = Azure::Blob::BlobService.new
+    # this is the reader # for this specific instance.
+    @reader = SecureRandom.uuid
```

Comment:
As you point out already, approaches 1 & 2 are not ideal either. For 2nd approach, handling multiple clients trying to create the same lock file without conflict already sounds like a disaster.

There are some other not ideal approaches:

3. Allow user to set the readerid manually, so that the new client can take over whatever crashed. The defect would be user interactive needed and specifying duplicated reader id is going to break the mutex.
4. Asking user to set the reader in the file to null manually. Also, requires user interactive, and may leads to invalid registry file.

That said, I don't see any of those trivial implementation.

And according the document: [stop is called outside of the plugin thread](https://www.elastic.co/guide/en/logstash/current/contributing-to-logstash.html).

And we also follow the example from the document for shutting down API: [sample code for the plugin shutdown APIs is available](https://github.com/logstash-plugins/logstash-input-example/blob/master/lib/logstash/inputs/example.rb)

My suggest on this is:
Open an issue to gathering feedback. If that is a real scenario, we then spend effort to polish that scenario.
What do you say?